### PR TITLE
Toponaming: transfer in getElementName and dependencies

### PR DESCRIPTION
**Pull request by bgbsww**

### CODE REVIEW

**Comment by Pesc0** [comment]
I am pretty sure findtaginelementname has already been moved in... Somewhere. Perhaps in elementnamingutils? I dont remember

**Comment by chennes** [comment]
> I am pretty sure findtaginelementname has already been moved in... Somewhere. Perhaps in elementnamingutils? I dont remember

Should be in MappedName.[h/cpp]

File: src/App/IndexedName.h [hunk 1]
```diff
@@ -147,6 +147,18 @@ class AppExport IndexedName {
         return result;
     }
 
+    const char * toString(std::string & s) const
```
Comment (chennes) [inline src/App/IndexedName.h:150]: My intent was that this be replaced by `appendToStringBuffer` -- do we need the char * return someplace? (Sorry, this was early days of the project, documentation is not what it should be)

File: src/App/MappedName.h [hunk 2]
```diff
@@ -474,6 +474,28 @@ class AppExport MappedName
         return appendToBuffer(res, startPosition, len);
     }
 
+    const char * toString(std::string &s, int from=0, int len=-1) const
```
Comment (chennes) [inline src/App/MappedName.h:477]: Needs a better name, and a clang-tidy pass.

File: src/Mod/Part/App/PartFeature.cpp [hunk 3]
```diff
@@ -757,3 +757,235 @@ bool Part::checkIntersection(const TopoDS_Shape& first, const TopoDS_Shape& seco
...
+                    idxName.setIndex(ancestors.front());
+                    mapped.index = idxName;
+                    return App::GeoFeature::_getElementName(name, mapped);
+                }
+            }
+        }
+    }
+
+    return App::GeoFeature::getElementName(name, type);
+}
```
Comment (chennes) [inline src/Mod/Part/App/PartFeature.cpp:991]: Missing newline

File: src/App/MappedName.h (hunk 2 above)
Comment (bgbsww) [reply src/App/MappedName.h:477]: What if we call it appendToBuffer and then note that the method right below it already exists?  Ooops.

File: src/App/IndexedName.h (hunk 1 above)
Comment (bgbsww) [reply src/App/IndexedName.h:150]: Nope, I was being thorough after seeing the use in MappedName.  Missed the forest for the trees.  This is an object lesson in not letting an old semi-complete PR hang around.  I knew so little back when this was first written.

File: src/Mod/Part/App/PartFeature.cpp [hunk 4]
```diff
@@ -757,3 +757,235 @@ bool Part::checkIntersection(const TopoDS_Shape& first, const TopoDS_Shape& seco
...
+        //
+        // In theory, all it takes to find one lower element that only appear
+        // in the given higher element. To make the algorithm more robust
+        // against model changes, we shall take minimum MinLowerTopoNames lower
+        // elements.
+        //
+        // On the other hand, it may be possible to take too many elements for
+        // disambiguation. We shall limit to maximum MaxLowerTopoNames. If the
+        // chosen elements are not enough to disambiguate the higher element,
+        // we'll include an index for disambiguation.
```
Comment (chennes) [inline src/Mod/Part/App/PartFeature.cpp:804]: Can you add rewriting this comment to you TODO list? Not for this PR, but as future cleanup. It's pretty hard to parse.

File: src/Mod/Part/App/TopoShape.h [hunk 5]
```diff
@@ -1131,13 +1131,19 @@ class PartExport TopoShape: public Data::ComplexGeoData
     static const std::string& shapeName(TopAbs_ShapeEnum type, bool silent = false);
     const std::string& shapeName(bool silent = false) const;
     static std::pair<TopAbs_ShapeEnum, int> shapeTypeAndIndex(const char* name);
+    static std::pair<TopAbs_ShapeEnum,int> shapeTypeAndIndex(const Data::IndexedName &name);
```
Comment (chennes) [inline src/Mod/Part/App/TopoShape.h:1134]:
```suggestion
    static std::pair<TopAbs_ShapeEnum, int> shapeTypeAndIndex(const Data::IndexedName &name);
```

File: src/Mod/Part/App/TopoShape.cpp [hunk 6]
```diff
@@ -477,6 +477,20 @@ std::pair<TopAbs_ShapeEnum,int> TopoShape::shapeTypeAndIndex(const char *name) {
...
+        return std::make_pair(TopAbs_SHAPE, 0);
+    static const std::string _subshape("SubShape");
+    if (boost::equals(element.getType(), _subshape))
+        return std::make_pair(TopAbs_SHAPE, element.getIndex());
+    TopAbs_ShapeEnum shapetype = shapeType(element.getType(), true);
+    if (shapetype == TopAbs_SHAPE)
+        return std::make_pair(TopAbs_SHAPE, 0);
+    return std::make_pair(shapetype, element.getIndex());
+}
+
```
Comment (chennes) [inline src/Mod/Part/App/TopoShape.cpp:493]:
```suggestion
std::pair<TopAbs_ShapeEnum, int> TopoShape::shapeTypeAndIndex(const Data::IndexedName& element)
{
    if (!element) {
        return std::make_pair(TopAbs_SHAPE, 0);
    }
    static const std::string _subshape("SubShape");
    if (boost::equals(element.getType(), _subshape)) {
        return std::make_pair(TopAbs_SHAPE, element.getIndex());
    }
    TopAbs_ShapeEnum shapetype = shapeType(element.getType(), true);
    if (shapetype == TopAbs_SHAPE) {
        return std::make_pair(TopAbs_SHAPE, 0);
    }
    return std::make_pair(shapetype, element.getIndex());
}

```

File: src/App/GeoFeature.cpp [hunk 7]
```diff
@@ -78,17 +79,46 @@ PyObject* GeoFeature::getPyObject()
...
     std::pair<std::string,std::string> ret;
     if(!name)
         return ret;
+    auto prop = getPropertyOfGeometry();
+    if(!prop) return std::make_pair("", name);
+
+    auto geo = prop->getComplexData();
+    if(!geo) return std::make_pair("", name);
+
+    return _getElementName(name, geo->getElementName(name));
```
Comment (chennes) [inline src/App/GeoFeature.cpp:96]:
```suggestion
    auto prop = getPropertyOfGeometry();
    if (!prop) {
        return std::make_pair("", name);
    }

    auto geo = prop->getComplexData();
    if (!geo) {
        return std::make_pair("", name);
    }

    return _getElementName(name, geo->getElementName(name));
```

File: src/Mod/Part/App/PartFeature.cpp (hunk 4 above)
Comment (bgbsww) [reply src/Mod/Part/App/PartFeature.cpp:804]: Done.

File: src/App/GeoFeature.cpp [hunk 8]
```diff
@@ -78,17 +79,46 @@ PyObject* GeoFeature::getPyObject()
...
+        const char *dot = strrchr(name,'.');
+        if(dot) {
+            // deliberately mangle the old style element name to signal a
+            // missing reference
+            ret.second = Data::MISSING_PREFIX;
+            ret.second += dot+1;
+        }
+    } else {
+        mapped.index.appendToStringBuffer(ret.second);
+    }
```
Comment (chennes) [inline src/App/GeoFeature.cpp:121]:
```suggestion
std::pair<std::string, std::string>
GeoFeature::_getElementName(const char* name, const Data::MappedElement& mapped) const
{
    std::pair<std::string, std::string> ret;
    if (mapped.index && mapped.name) {
        std::ostringstream ss;
        ss << Data::ComplexGeoData::elementMapPrefix() << mapped.name << '.' << mapped.index;
        ret.first = ss.str();
        mapped.index.appendToStringBuffer(ret.second);
    }
    else if (mapped.name) {
        //        FC_TRACE("element mapped name " << name << " not found in " << getFullName());
        ret.first = name;
        const char* dot = strrchr(name, '.');
        if (dot) {
            // deliberately mangle the old style element name to signal a
            // missing reference
            ret.second = Data::MISSING_PREFIX;
            ret.second += dot + 1;
        }
    }
    else {
        mapped.index.appendToStringBuffer(ret.second);
    }

```

**Review by chennes — APPROVED** [review]
All formatting changes, I'll take care of merging them.
